src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+                    if (!string.IsNullOrEmpty(scopeId))
+                    {
+                        if (scopeId.Length < 2)
+                        {
+                            scope = 0;
+                            return false;
+                        }
+
+                        for (int i = 1; i < scopeId.Length; i++)
+                        {
+                            char c = scopeId[i];
+                            if (c < '0' || c > '9')
+                            {
+                                scope = 0;
+                                return false;
+                            }
+                            result = (result * 10) + (c - '0');
+                            if (result > uint.MaxValue)
+                            {
+                                scope = 0;
+                                return false;
+                            }
+                        }
+                    }
+
+                    scope = (uint)result;
+                    return true;
+                }
+            }
+
+            scope = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Appends each of the numbers in address in indexed range [fromInclusive, toExclusive),
+        /// while also replacing the longest sequence of 0s found in that range with "::", as long
+        /// as the sequence is more than one 0.
+        /// </summary>
+        private static void AppendSections(ushort[] address, int fromInclusive, int toExclusive, StringBuilder buffer)
+        {
+            // Find the longest sequence of zeros to be combined into a "::"
+            (int zeroStart, int zeroEnd) = IPv6AddressHelper.FindCompressionRange(address, fromInclusive, toExclusive);
+            bool needsColon = false;
+
+            // Output all of the numbers before the zero sequence
+            for (int i = fromInclusive; i < zeroStart; i++)
+            {
+                if (needsColon)
+                {
+                    buffer.Append(':');
+                }
+                needsColon = true;
+                AppendHex(address[i], buffer);
+            }
+
+            // Output the zero sequence if there is one
+            if (zeroStart >= 0)
+            {
+                buffer.Append("::");
+                needsColon = false;
+                fromInclusive = zeroEnd;
+            }
+
+            // Output everything after the zero sequence
+            for (int i = fromInclusive; i < toExclusive; i++)
+            {
+                if (needsColon)
+                {
+                    buffer.Append(':');
+                }
+                needsColon = true;
+                AppendHex(address[i], buffer);
             }
         }
+
+        /// <summary>Appends a number as hexadecimal (without the leading "0x") to the StringBuilder.</summary>
+        private static unsafe void AppendHex(int value, StringBuilder buffer)
+        {
+            const int MaxLength = 8;
```

Comment:
It looks like this is always called with a short.  Is that right?  Should value be a short and MaxLength = 4?